transport/tlsfrag/writer.go

```diff
@@ -0,0 +1,179 @@
+	}
+	if frag == nil {
+		return nil, errors.New("frag callback function must not be nil")
+	}
+	fw := &clientHelloFragWriter{
+		base: base,
+		frag: frag,
+		buf:  newClientHelloBuffer(),
+	}
+	if rf, ok := base.(io.ReaderFrom); ok {
+		return &clientHelloFragReaderFrom{fw, rf}, nil
+	}
+	return fw, nil
+}
+
+// Write implements io.Writer.Write. It attempts to split the data received in the first one or more Write call(s)
+// into two TLS records if the data corresponds to a TLS Client Hello record.
+func (w *clientHelloFragWriter) Write(p []byte) (n int, err error) {
+	if w.done {
+		return w.base.Write(p)
+	}
+	if w.rcds != nil {
+		if _, err = w.flushRecords(); err != nil {
+			return
+		}
+		return w.base.Write(p)
+	}
+
+	if n, err = w.buf.Write(p); err != nil {
+		if errors.Is(err, errTLSClientHelloFullyReceived) {
+			w.splitBufToRecords()
+		} else {
+			w.copyBufToRecords()
+		}
+		// We did not call w.Write(p[n:]) here because p[n:] might be empty, and we don't want to
+		// Write an empty buffer to w.base if it's not initiated by the upstream caller.
+		if _, err = w.flushRecords(); err != nil {
+			return
+		}
+		if p = p[n:]; len(p) > 0 {
+			m, e := w.base.Write(p)
+			n += m
+			err = e
+		}
+		return
+	}
+
+	if n < len(p) {
+		return n, io.ErrShortWrite
+	}
+	return
+}
+
+// ReadFrom implements io.ReaderFrom.ReadFrom. It attempts to split the first packet into two TLS records if the data
+// corresponds to a TLS Client Hello record. And then copies the remaining data from r to the base io.Writer until EOF
+// or error.
+//
+// If the first packet is not a valid TLS Client Hello, everything from r gets copied to the base io.Writer as is.
+//
+// It returns the number of bytes read. Any error except EOF encountered during the read is also returned.
+func (w *clientHelloFragReaderFrom) ReadFrom(r io.Reader) (n int64, err error) {
+	if w.done {
+		return w.baseRF.ReadFrom(r)
+	}
+	if w.rcds != nil {
+		if _, err = w.flushRecords(); err != nil {
+			return
+		}
+		return w.baseRF.ReadFrom(r)
+	}
+
+	if n, err = w.buf.ReadFrom(r); err != nil {
+		if errors.Is(err, errTLSClientHelloFullyReceived) {
+			w.splitBufToRecords()
+		} else {
+			w.copyBufToRecords()
+		}
+		// recursively flush w.rcds and read the remaining content from r
+		m, e := w.ReadFrom(r)
+		return n + m, e
+	}
+	return
+}
+
+// copyBuf copies w.buf into w.rcds.
+func (w *clientHelloFragWriter) copyBufToRecords() {
+	w.rcds = bytes.NewBuffer(w.buf.Bytes())
+	w.buf = nil // allows the GC to recycle the memory
+}
+
+// splitBuf splits w.buf into two records and put them into w.rcds.
+func (w *clientHelloFragWriter) splitBufToRecords() {
+	content := w.buf.Bytes()[recordHeaderLen:]
+	split := w.frag(content)
+	if split <= 0 || split >= len(content) {
+		w.copyBufToRecords()
+		return
+	}
+
+	header := make([]byte, recordHeaderLen)
+	copy(header, w.buf.Bytes())
+
+	w.rcds = bytes.NewBuffer(make([]byte, 0, w.buf.Len()+recordHeaderLen))
```

Comment:
Please remove this allocation and copy. You can write directly from the record buffer:
1. update record len for head
2. shift tail and add a header for it with proper length
3. write the full buffer

This requires the buffer to have recordLength + 2 * recordHeaderLen bytes

Reply:
Yeah, but there is a tradeoff here, we either need to expose implementation details in `buffer.go` (with an extra 5 bytes padding that is not used anywhere in `buffer.go`); if we'd like to keep the type doing what it should do, we need to allocate additional space.